src/modules/csm/csm.py

```diff
@@ -61,22 +61,41 @@ def build_report(self, blockstamp: ReferenceBlockStamp) -> tuple:
         last_ref_slot = self.w3.csm.get_csm_last_processing_ref_slot(blockstamp)
         ref_slot = self.get_current_frame(blockstamp).ref_slot
 
-        # Get module's node operators.
-        _ = self.module_validators_by_node_operators(blockstamp)
-        # Read performance threshold value from somewhere (hardcoded?).
-        _ = self.frame_performance.avg_perf * 0.95
-        # Build the map of the current distribution operators.
-        # _ = groupby(self.frame_performance.aggr_per_val, operators)
-        # Exclude validators of operators with stuck keys.
-        _ = self.w3.csm.get_csm_stuck_node_operators(
+        threshold = self.frame_performance.avg_perf * self.w3.csm.oracle.perf_threshold(blockstamp.block_hash)
+        stuck_operators = self.w3.csm.get_csm_stuck_node_operators(
             self._slot_to_block_identifier(last_ref_slot),
             self._slot_to_block_identifier(ref_slot),
         )
-        # Exclude underperforming validators.
+
+        operators = self.module_validators_by_node_operators(blockstamp)
+        # Build the map of the current distribution operators.
+        distribution: dict[NodeOperatorId, int] = {}
+        total = 0
+
+        for (_, no_id), validators in operators.items():
+            if no_id in stuck_operators:
+                continue
+
+            share = len(
+                [
+                    v
+                    for v in validators
+                    if self.frame_performance.perf(ValidatorIndex(int(v.index))) > threshold
+                ]
+            )
+
+            distribution[no_id] = share
+            total += share
 
         # Calculate share of each CSM node operator.
-        _ = self._to_distribute(blockstamp)
-        shares: tuple[tuple[NodeOperatorId, int]] = tuple()  # type: ignore
+        to_distribute = self.w3.csm.fee_distributor.pending_to_distribute(blockstamp.block_hash)
+        shares: list[tuple[NodeOperatorId, int]] = []
+        for no_id, share in distribution.items():
+            shares.append((no_id, to_distribute * share // total))
+
+        distributed = sum((s for (_, s) in shares))
```

Comment:
`total_distributed` and having increment in the cycle above?

Reply:
Looks better in this way.